Add support for Framework SD Expansion Card

JohnAZoidberg opened the pull request:
It was just recently [announced](https://frame.work/tw/en/blog/introducing-the-new-framework-laptop-13-with-intel-core-ultra-series-1-processors) together with the Intel Core Ultra platforms.
It will be available soon here: https://frame.work/tw/en/products/sd-expansion-card

It uses a GL3230 controller, so it's already supported by the plugin, but the plugin currently does not handle non-genesys VID/PID devices.

hughsie commented on plugins/genesys-gl32xx/fu-genesys-gl32xx-device.c:
```diff
@@ -912,11 +918,38 @@ fu_genesys_gl32xx_device_set_progress(FuDevice *self, FuProgress *progress)
 	self->packetsz = 64;
-	fu_device_set_vendor(FU_DEVICE(self), "Genesys");
+	self->compatible_model = 0;
+	if (fu_device_has_vendor_id(FU_DEVICE(self), "USB:05E3")) {
+		fu_device_set_vendor(FU_DEVICE(self), "Genesys");
+	}
```
this you can do with:

    [USB\VID_05E3]
    Vendor = Genesys

JohnAZoidberg replied:
In the quirk file? Not something like `[BLOCK\VEN_05E3]` ? What's the difference between those formats?

But actually we might be able to use the default name fwupd gets, let me check I think we have a firmware that makes it looks like a generic genesys device.

JohnAZoidberg replied:
This is what it looks like without quirk:

```
> sudo ./venv/bin/fwupdtool get-devices --plugins genesys_gl32xx
Place your right index finger on the fingerprint reader
Loading…                 [************************************** ]
WARNING: This package has not been validated, it may not work properly.
Framework Laptop (12th Gen Intel Core)
│
└─MassStorageClass:
      Device ID:          7818afbec6007ce4fb6e2792851f6e46eefb7d85
      Current version:    2907
      Vendor:             Generic (BLOCK:0x05E3)
      GUID:               0cc25308-a257-5fe9-8183-abad7b85b050 ← BLOCK\VEN_05E3&DEV_0764&VER_29&CID_E4F090C2
      Device Flags:       • Updatable
                          • Is in bootloader mode
                          • Cryptographic hash verification is available
                          • Unsigned Payload
```

Coming from the SCSI strings:

```
> lsscsi --no-nvme
[1:0:0:0]    disk    Generic  MassStorageClass 2907  /dev/sdb 
```

For reference, the USB descriptor looks like this:

```
> lsusb -d 05e3:
Bus 002 Device 017: ID 05e3:0764 Genesys Logic, Inc. USB Storage
```

So we should still do the override the name.

hughsie commented on plugins/genesys-gl32xx/genesys-gl32xx.quirk:
```diff
@@ -5,3 +5,11 @@ FirmwareSize = 0x10000
+Name = Framework SD Expansion Card
+Vendor = Framework
+# Compatible with Genesys PID 0x0764
+CompatibleModel = 0x0764
```
This needs to be the slightly verbose `GenesysGl32xxCompatibleModel` and you need to do `fu_context_add_quirk_key(ctx, "GenesysGl32xxCompatibleModel");` in the plugin.c too -- otherwise the daemon will get mad at you.

hughsie commented on plugins/genesys-gl32xx/fu-genesys-gl32xx-device.c:
```diff
@@ -681,10 +685,12 @@ fu_genesys_gl32xx_device_setup(FuDevice *device, GError **error)
-			       self->chip_name,
-			       fu_udev_device_get_model(FU_UDEV_DEVICE(device)));
-	fu_device_set_name(device, name);
+	if (fu_device_has_vendor_id(device, "USB:05E3")) {
```
you might be able to do this using `fu_device_get_name(device) == NULL`

JohnAZoidberg replied:
I tried, but that gave me the name from SCSI inquiry.

JohnAZoidberg replied:
Whoops, this didn't work. Tried it with genesys vid firmware and I had to change the condition to `BLOCK:0x05E3`.

hughsie commented on plugins/genesys-gl32xx/fu-genesys-gl32xx-device.c:
```diff
@@ -488,9 +489,12 @@ fu_genesys_gl32xx_device_ensure_cid(FuGenesysGl32xxDevice *self, GError **error)
+	guint16 model = fu_udev_device_get_model(FU_UDEV_DEVICE(self));
 	guint8 data[4] = {0};
 
+	if (self->compatible_model)
```
`!= 0` pls

hughsie commented on plugins/genesys-gl32xx/fu-genesys-gl32xx-device.c:
```diff
@@ -912,11 +918,35 @@ fu_genesys_gl32xx_device_set_progress(FuDevice *self, FuProgress *progress)
 {
 	self->packetsz = 64;
-	fu_device_set_vendor(FU_DEVICE(self), "Genesys");
+	self->compatible_model = 0;
```
You don't need this -- it's all zero cleared.

hughsie reviewed (commented):
A tiny nit and then LGTM.

JohnAZoidberg commented:
> A tiny nit and then LGTM.

Fixed and squashed fixup commits.

I have a question, you seem to merge PRs into master but releases are done from `1_9_X` branch.
Do you have documentation about when and how commits are ported to a release branch?

superm1 commented:
> I have a question, you seem to merge PRs into master but releases are done from 1_9_X branch.

This is just because we've got an ABI break in main right now and want to get everything right before we commit to the new ABI for a number of years.  I think after Richard finishes up the new uevent device provider we'll probably be ready for a 2.0.0 release.

> Do you have documentation about when and how commits are ported to a release branch?

Basically when it's in main, checkout 1_9_X as another branch and do a `git cherry-pick -x SHA` where SHA is the commit you want.  Submit a PR and it will run through CI and then it can merge to next 1.9.x. release.

JohnAZoidberg commented:
Clarification, the controller is GL3232S not GL3230. For anybody finding this PR in the future - to avoid confusion.
